File: leetcode/738.monotone-increasing-digits/go/monotoneincreasingdigits.go
package monotoneincreasingdigits

import "strconv"

func monotoneIncreasingDigits(n int) int {
	arr := []byte(strconv.Itoa(n))
	marker := len(arr)

	for i := len(arr) - 1; i > 0; i-- {
		if arr[i-1] > arr[i] {
			arr[i-1]--
			marker = i
		}
	}

	for i := marker; i < len(arr); i++ {
		arr[i] = '9'
	}

	res, _ := strconv.Atoi(string(arr))
	return res
}

File: leetcode/738.monotone-increasing-digits/ts/monotoneincreasingdigits.ts
const monotoneIncreasingDigits = (n: number): number => {
    const arr: string[] = n.toString().split("")
    let marker = arr.length

    for (let i = arr.length - 1; i > 0; i--) {
        if (arr[i - 1] > arr[i]) {
            arr[i - 1] = String.fromCharCode(arr[i - 1].charCodeAt(0) - 1)
            marker = i
        }
    }

    for (let i = marker; i < arr.length; i++) {
        arr[i] = "9"
    }

    return Number(arr.join(""))
}
